### ISSUE TITLE
Forecast Microservice: fribbe-forecast-service

### ISSUE THREAD

**Issue by FlorianObermayer** [FlorianObermayer/fribbe-status-checker#38]
# Forecast Microservice: fribbe-forecast-service

## Overview

A stateless Python microservice that accepts pre-aggregated feature data and returns occupancy predictions. No direct DB access - the main app is the data broker. Designed to run on a Raspberry Pi 4 (8 GB RAM) alongside the main app, or to be outsourced to any other host by changing `FORECAST_SERVICE_URL` in `.env`.

---

## Project layout

```
fribbe-forecast-service/
├── Dockerfile
├── pyproject.toml
├── app/
│   ├── main.py             # FastAPI app
│   ├── model.py            # inference + training logic
│   └── schemas.py          # request/response Pydantic models
├── tests/
│   └── ...
└── artifacts/              # persisted via Docker volume
    └── model.joblib        # trained model artifact
```

---

### `POST /forecast/calculate`

```json
// Request
{
  "target_dates": ["2025-07-15"],
  "features": {
    "historic_occupancy": [{"timestamp": "...", "count": 4}],
    "weather_forecast":   [{"timestamp": "...", "temp_c": 28.5, "precipitation_mm": 0.0, "wind_kph": 12, "cloud_cover_pct": 20}],
    "tournaments":        [{"date": "...", "location_type": "on_site", "participant_count": 32}],
    "context":            {"public_holidays": ["2025-07-15"], "season": "summer"}
  }
}

// Response
{
  "forecasts": [
    {
      "date": "2025-07-15",
      "buckets": [
        {"bucket": "morning",   "level": "medium", "confidence": 0.74},
        {"bucket": "mid_day",   "level": "high",   "confidence": 0.81},
        {"bucket": "afternoon", "level": "high",   "confidence": 0.79},
        {"bucket": "evening",   "level": "low",    "confidence": 0.65},
        {"bucket": "night",     "level": "low",    "confidence": 0.90}
      ]
    }
  ]
}
```

### `POST /forecast/train`

Same feature format as `/calculate` but with labelled training data (historic occupancy counts per bucket). Returns model version + basic metrics (e.g. accuracy, F1 per class).

```json
// Response
{
  "model_version": "2025-07-15T05:00:00Z",
  "metrics": {
    "accuracy": 0.78,
    "f1_per_class": {"low": 0.82, "medium": 0.71, "high": 0.76}
  }
}
```

### `GET /health`

Returns service status, model version, last training timestamp. Returns HTTP 503 if no model artifact exists (cold start).

---

## Tasks

- [ ] Create `fribbe-forecast-service/` repo/directory with `Dockerfile` and `pyproject.toml`
- [ ] FastAPI app with bearer token auth (`Authorization: Bearer <FORECAST_API_TOKEN>`) on all endpoints except `/health`
- [ ] `POST /forecast/calculate` - run inference, return predictions per bucket; return 503 if no trained model exists
- [ ] `POST /forecast/train` - retrain model, persist artifact to `artifacts/`, return version + metrics
- [ ] `GET /health` - report model version, last training timestamp, readiness
- [ ] Feature engineering: map hourly records to 5 time buckets (#36 defines the canonical bucket boundaries), derive `day_of_week`, `is_weekend`, `is_holiday`, `season`
- [ ] Model: **LightGBM** classifier. See model selection notes below.
- [ ] Handle cold start: return HTTP 503 with clear error message when no model artifact exists; `/health` reports `"ready": false`
- [ ] Handle missing/partial features gracefully (fallback to `low` with low confidence)
- [ ] Store artifact as `joblib` file in a Docker volume
- [ ] Add unit tests for feature engineering and prediction with a fixture model
- [ ] Multi-arch Dockerfile (`linux/arm64` + `linux/amd64`) for Pi compatibility

---

## Model selection notes

- **LightGBM** (4.6.0+) ships prebuilt ARM64 wheels on PyPI (`py3-none-manylinux2014_aarch64`, 3.5 MB). No CMake or build-from-source needed on Raspberry Pi.
- LightGBM handles categorical features natively (`day_of_week`, `season`, `bucket`) - no one-hot encoding needed. It also handles missing values natively, which simplifies feature engineering when weather data has gaps.
- With the expected data volume (a few hundred to low thousands of training rows), accuracy will be comparable to scikit-learn. The advantage is simpler preprocessing code, not raw performance.
- Fallback: if LightGBM causes issues on ARM64 for any reason, swap to scikit-learn `GradientBoostingClassifier` - the microservice API stays the same.
- **Do not use** PyTorch, TensorFlow, or large transformer models. They far exceed the Pi's memory budget.

## Resource budget

- Target: **< 512 MB RSS** during inference, **< 1.5 GB peak** during training
- The main app + this service + OS should fit comfortably in the Pi's 8 GB with room for filesystem cache
